lib/utils/utf8.c

```diff
@@ -79,3 +87,40 @@ char *utf8_lcpy(char *dst, const char *src, size_t n)
 
 	return dst;
 }
+
+bool utf8_is_valid(const unsigned char *str, size_t maxlen)
+{
+	size_t i = 0, nbyte = 0;
+	size_t len;
+
+	/* It will also return false */
+	if (str == NULL) {
+		return false;
+	}
+	len = strnlen(str, maxlen);
```

Comment:
I'm sorry, but I fail to see the need to have `strnlen` or `strlen` here. Worst case scenario is that we loop the entire string twice, without any benefit. If the first two characters aren't valid utf8, we could already stop after 2 steps.

Reply:
Good proposal. Maybe I can use len as a parameter. I won't deal with the exception caused by the wrong len. WDYT?

Reply:
I will modify it later and change `i < len` to `i < maxlen`.

Reply:
> If the first two characters aren't valid utf8, we could already stop after 2 steps.

```
if (str[i] <= ASCII_CHAR && str[i] >= '\0') {
			i++;
			continue;
		} else {
			if (str[i] <= SEQUENCE_MAX_LEN_2_BYTE
			 && str[i] >= SEQUENCE_MIN_LEN_2_BYTE) {
				nbyte = 2;
			} else if (str[i] <= SEQUENCE_MAX_LEN_3_BYTE
					&& str[i] >= SEQUENCE_MIN_LEN_3_BYTE) {
				nbyte = 3;
			} else if (str[i] <= SEQUENCE_MAX_LEN_4_BYTE
					&& str[i] >= SEQUENCE_MIN_LEN_4_BYTE) {
				nbyte = 4;
			} else {
				return false;
			}
		}
```
I almost forgot, my implementation includes this function, it is in the recognition. It will exit the function when it matches non-UTF8. The same is true for the first character.

Reply:
And, I have another question. You mentioned that you can put maxlen directly into the loop instead of using strnlen, which seems to be no different from directly using len as a function parameter. In fact, it can be len directly instead of maxlen. Right?
@pdgendt